test/tapeimage.cpp

```diff
@@ -385,6 +385,67 @@ TEST_CASE(
     lfp_close(tif);
 }
 
+TEST_CASE(
+    "Peek into the underlying protocol"
+    "[tapeimage][peek]") {
+
+    const auto file = std::vector< unsigned char > {
+        0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        0x14, 0x00, 0x00, 0x00,
+
+        /* begin body */
+        0x01, 0x02, 0x03, 0x04,
+        0x05, 0x06, 0x07, 0x08,
+        /* end body */
+
+        0x01, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        0x20, 0x00, 0x00, 0x00,
+
+        0x01, 0x00, 0x00, 0x00,
+        0x14, 0x00, 0x00, 0x00,
+        0x2C, 0x00, 0x00, 0x00,
+    };
+
+    const auto expected1 = std::vector< unsigned char > {
+        0x01, 0x02, 0x03, 0x04,
+    };
+
+    const auto expected2 = std::vector< unsigned char > {
+        0x05, 0x06, 0x07, 0x08,
+    };
+
+    auto* mem = lfp_memfile_openwith(file.data(), file.size());
+    auto* tif = lfp_tapeimage_open(mem);
+
+    auto out = std::vector< unsigned char >(4, 0xFF);
+    std::int64_t bytes_read = -1;
+    auto err = lfp_readinto(tif, out.data(), 4, &bytes_read);
+
+    CHECK(bytes_read == 4);
+    CHECK(err == LFP_OK);
+    CHECK_THAT(out, Equals(expected1));
+
+    std::int64_t tif_tell;
+    lfp_tell(tif, &tif_tell);
+    CHECK(tif_tell == 4);
+
+    lfp_protocol* flesh;
+    lfp_peek(tif, &flesh);
+
+    std::int64_t mem_tell;
+    lfp_tell(flesh, &mem_tell);
+    CHECK(mem_tell == 16);
+
+    err = lfp_readinto(tif, out.data(), 4, &bytes_read);
```

Comment:
I don't think you should break your promise and do any non-constant operation here - you break your outer protocol tif (and you still want to close it in the end)